feature/gribi/mplsutil/mplsutil.go

```diff
@@ -0,0 +1,474 @@
+	// is non-nil it is called after the gRIBI programming is verified.
+	//
+	// The DUT is expected to be in a topology where 192.0.2.2 resolves to
+	// a valid next-hop.
+	PopNLabels
+	// PopOnePushN implements a test whereby one (the top) label is popped, and N labels as specified
+	// by pushLabels are pushed to the stack for an input MPLS packet. Two LFIB entries (100 and 200)
+	// are created. If trafficFunc is non-nil it is called after the gRIBI programming has been validated.
+	//
+	// The DUT is expected to be in a topology where 192.0.2.2 is a resolvable next-hop.
+	PopOnePushN
+)
+
+type GRIBIMPLSTest struct {
+	// mode stores the mode that the test was initialised in.
+	mode Mode
+
+	// defaultNIName stores the name of the default network instance for the
+	// test.
+	defaultNIName string
+
+	// client stores the gRIBI client that should be used to contact the
+	// DUT.
+	client *fluent.GRIBIClient
+
+	// config is a cache of the OTG configuration that is to be used for
+	// the test.
+	otgConfig gosnappi.Config
+
+	// args stores the arguments to the test.
+	args *Args
+
+	// result caches the last set of results that the test ran.
+	result []*client.OpResult
+}
+
+func New(c *fluent.GRIBIClient, m Mode, defName string, args *Args) *GRIBIMPLSTest {
+	return &GRIBIMPLSTest{
+		mode:          m,
+		defaultNIName: defName,
+		args:          args,
+	}
+}
+
+type Args struct {
+	// LabelsToPop specifies the set of labels that should be popped from
+	// an incoming MPLS packet.
+	LabelsToPop []uint32
+	// LabelsToPush specifies the set of labels that should be pushed to
+	// an incoming MPLS packet.
+	LabelsToPush []uint32
+}
+
+// ConfigureDevices configures the DUT and ATE with their base configurations
+// such the test topology is set up. This uses a constant topology as described
+// in topo.go.
+func (g *GRIBIMPLSTest) ConfigureDevices(t *testing.T, dut *ondatra.DUTDevice, ate *ondatra.ATEDevice) {
+	DUTSrc.Name = dut.Port(t, "port1").Name()
+	DUTDst.Name = dut.Port(t, "port2").Name()
+
+	otgCfg, err := configureATEInterfaces(t, ate, ATESrc, DUTSrc, ATEDst, DUTDst)
+	if err != nil {
+		t.Fatalf("cannot configure ATE interfaces via OTG, %v", err)
+	}
+
+	ate.OTG().PushConfig(t, otgCfg)
+
+	d := &oc.Root{}
+	// configure ports on the DUT. note that each port maps to two interfaces
+	// because we create a LAG.
+	for index, i := range []*attrs.Attributes{DUTSrc, DUTDst} {
+		cfgs, err := dutIntf(dut, i, index)
+		if err != nil {
+			t.Fatalf("cannot generate configuration for interface %s, err: %v", i.Name, err)
+		}
+		for _, intf := range cfgs {
+			d.AppendInterface(intf)
+		}
+	}
+	fptest.LogQuery(t, "", gnmi.OC().Config(), d)
+	gnmi.Update(t, dut, gnmi.OC().Config(), d)
+
+	// Sleep for 1 second to ensure that OTG has absorbed configuration.
+	time.Sleep(1 * time.Second)
+	ate.OTG().StartProtocols(t)
+
+	g.otgConfig = otgCfg
+}
+
+func (g *GRIBIMPLSTest) ProgramGRIBI(t *testing.T) {
```

Comment:
I think we have some inconsistency here -- since `ConfigureDevices` also touches the DUT. It's probably useful to go with the rule that if it touches the DUT it isn't a helper to get the most detailed debugging information.